pkg/collector/autodiscovery/configresolver.go

```diff
@@ -135,20 +135,28 @@ func (cr *ConfigResolver) ResolveTemplate(tpl check.Config) []check.Config {
 // resolve takes a template and a service and generates a config with
 // valid connection info and relevant tags.
 func (cr *ConfigResolver) resolve(tpl check.Config, svc listeners.Service) (check.Config, error) {
-	vars := tpl.GetTemplateVariables()
-	for _, v := range vars {
-		name, key := parseTemplateVar(v)
-		if f, ok := templateVariables[string(name)]; ok {
-			resolvedVar := f(key, svc)
-			if resolvedVar != nil {
-				tpl.InitConfig = bytes.Replace(tpl.InitConfig, v, resolvedVar, -1)
-				tpl.Instances[0] = bytes.Replace(tpl.Instances[0], v, resolvedVar, -1)
+	for i := 0; i < len(tpl.Instances); i++ {
+		vars := tpl.GetTemplateVariablesForInstance(i)
+		for _, v := range vars {
+			name, key := parseTemplateVar(v)
+			if f, ok := templateVariables[string(name)]; ok {
+				resolvedVar := f(key, svc)
+				if resolvedVar != nil {
+					// init config vars are replaced by the first found
+					tpl.InitConfig = bytes.Replace(tpl.InitConfig, v, resolvedVar, -1)
+					tpl.Instances[i] = bytes.Replace(tpl.Instances[i], v, resolvedVar, -1)
+				}
+			} else {
+				return check.Config{}, fmt.Errorf("template variable %s does not exist", name)
 			}
-		} else {
-			return check.Config{}, fmt.Errorf("template variable %s does not exist", name)
 		}
+		tags, err := svc.GetTags()
```

Comment:
since tags are not cached in the service anymore I think we should only call `svc.GetTags()` once in resolve. Otherwise we may have a race condition where tags get evicted from the tagger cache (expiry) and the last instance gets different tags from the first.

It's better to have all instances be consistent, although with outdated tags.

Reply:
Agreed, moving the `getTags` to the top of the method